src/Tasks/Microsoft.NET.Build.Tasks/CompileReadyToRunImages.cs

```diff
@@ -0,0 +1,415 @@
+        [Output]
+        public ITaskItem[] R2RFilesToPublishPreserveNewest => _r2rPublishPreserveNewest.ToArray();
+
+        protected override string ToolName => Path.GetFileName(_crossgenPath);
+        protected override string GenerateFullPathToTool() => _crossgenPath;
+
+        private List<ITaskItem> _r2rPublishAlways = new List<ITaskItem>();
+        private List<ITaskItem> _r2rPublishPreserveNewest = new List<ITaskItem>();
+
+        private string _runtimeIdentifier;
+        private string _packagePath;
+        private string _crossgenPath;
+        private string _clrjitPath;
+        private string _platformAssembliesPath;
+        private string _diasymreaderPath;
+        private string _pathSeparatorCharacter = ";";
+
+        private Architecture _targetArchitecture;
+
+
+        public override bool Execute()
+        {
+            // Get the list of runtime identifiers that we support and can target
+            ITaskItem frameworkRef = KnownFrameworkReferences.Where(item => String.Compare(item.ItemSpec, "Microsoft.NETCore.App", true) == 0).SingleOrDefault();
+            string supportedRuntimeIdentifiers = frameworkRef == null ? null : frameworkRef.GetMetadata("RuntimePackRuntimeIdentifiers");
+
+            // Get information on the runtime package used for the current target
+            ITaskItem frameworkPack = RuntimePacks.Where(pack => pack.ItemSpec.EndsWith(".Microsoft.NETCore.App", StringComparison.InvariantCultureIgnoreCase)).SingleOrDefault();
+            _runtimeIdentifier = frameworkPack == null ? null : frameworkPack.GetMetadata(MetadataKeys.RuntimeIdentifier);
+            _packagePath = frameworkPack == null ? null : frameworkPack.GetMetadata(MetadataKeys.PackageDirectory);
+
+            var runtimeGraph = new RuntimeGraphCache(BuildEngine4, Log).GetRuntimeGraph(RuntimeGraphPath);
+            var supportedRIDsList = supportedRuntimeIdentifiers == null ? Array.Empty<string>() : supportedRuntimeIdentifiers.Split(';');
+
+            // Get the best RID for the host machine, which will be used to validate that we can run crossgen for the target platform and architecture
+            string hostRuntimeIdentifier = NuGetUtils.GetBestMatchingRid(
+                runtimeGraph, 
+                DotNet.PlatformAbstractions.RuntimeEnvironment.GetRuntimeIdentifier(), 
+                supportedRIDsList, 
+                out bool wasInGraph);
+
+            if (hostRuntimeIdentifier == null || _runtimeIdentifier == null || _packagePath == null)
+            {
+                Log.LogError(Strings.ReadyToRunNoValidRuntimePackageError);
+                return false;
+            }
+
+            if (!ExtractTargetPlatformAndArchitecture(_runtimeIdentifier, out string targetPlatform, out _targetArchitecture) ||
+                !ExtractTargetPlatformAndArchitecture(hostRuntimeIdentifier, out string hostPlatform, out Architecture hostArchitecture) ||
+                targetPlatform != hostPlatform)
```

Comment:
Hmm, this still wouldn't work if we had say separate win-x64 and win10-x64 runtime packs in the future. If the host was win10-x64 and the target of the app is win7-x64, then we'd get to win != win10.

This isn't really the check I was wondering about when we chatted. I was wondering if instead we could check if the host rid is compatible with the target rid. I think that would be `runtimeGraph.ExpandRuntime(hostRid).Contains(targetRid)`. But as I mentioned that doesn't work either because the graph claims linux-musl can fallback to linux and also we use different architectures for host and 

Sigh, this is complicated. It's no fun to have a build tool that couples the target OS to the host OS like this. :(

Reply:
I think we can keep this check the way it is for now. In the future, if we decided to add a win10-x64 and win7-64 that do not fall back to win-x64, we can revisit this.
From a crossgen perspective, there is no difference between the different versions of Windows, and we'll need a graph that is able to fallback to just "win", and if that's not possible, we'll need to implement some logic here that detects "win", regardless of the version.
The call to GetBestMatchingRid right now handles all these cases, because it mirrors what we do for the target platform.